src/qibolab/instruments/rfsoc.py

```diff
@@ -1,54 +1,156 @@
 """ RFSoC FPGA driver.
 
-This driver needs the library Qick installed
-Supports the following FPGA:
+This driver needs a Qibosoq server installed and running
+Tested on the following FPGA:
  *   RFSoC 4x2
+ *   ZCU111
 """
 
-import pickle
+import json
 import socket
-from dataclasses import dataclass
-from typing import Dict, List, Optional, Tuple, Union
+from dataclasses import asdict
+from typing import Dict, List, Tuple, Union
 
 import numpy as np
+import qibosoq.abstracts as rfsoc
 
 from qibolab import AcquisitionType, AveragingMode, ExecutionParameters
 from qibolab.instruments.abstract import AbstractInstrument
 from qibolab.platforms.abstract import Qubit
-from qibolab.pulses import PulseSequence, PulseType
+from qibolab.pulses import Drag, Gaussian, Pulse, PulseSequence, PulseType, Rectangular
 from qibolab.result import IntegratedResults, SampleResults
 from qibolab.sweeper import Parameter, Sweeper
 
-
-@dataclass
-class QickProgramConfig:
-    sampling_rate: int = 9_830_400_000
-    repetition_duration: int = 100_000
-    adc_trig_offset: int = 200
-    max_gain: int = 32_000
-    reps: int = 1000
-    expts: Optional[int] = None
-
-
-class TII_RFSOC4x2(AbstractInstrument):
-    """Instrument object for controlling the RFSoC4x2 FPGA.
-    Playing pulses requires first the execution of the ``setup`` function.
+HZ_TO_MHZ = 1e-6
+NS_TO_US = 1e-3
+
+
+def convert_qubit(qubit: Qubit) -> rfsoc.Qubit:
+    """Convert `qibolab.platforms.abstract.Qubit` to `qibosoq.abstract.Qubit`"""
+    if qubit.flux:
+        return rfsoc.Qubit(qubit.flux.bias, qubit.flux.ports[0][1])
+    return rfsoc.Qubit(0.0, None)
+
+
+def convert_pulse(pulse: Pulse, qubits: Dict) -> rfsoc.Pulse:
+    """Convert `qibolab.pulses.pulse` to `qibosoq.abstract.Pulse`"""
+
+    pulse_type = pulse.type.name.lower()
+    dac = getattr(qubits[pulse.qubit], pulse_type).ports[0][1]
+    adc = qubits[pulse.qubit].feedback.ports[0][1] if pulse_type == "readout" else None
+
+    try:
+        lo_frequency = getattr(qubits[pulse.qubit], pulse_type).local_oscillator._frequency
+    except NotImplementedError:
+        lo_frequency = 0
+
+    rfsoc_pulse = rfsoc.Pulse(
+        frequency=(pulse.frequency - lo_frequency) * HZ_TO_MHZ,
+        amplitude=pulse.amplitude,
+        relative_phase=np.degrees(pulse.relative_phase),
+        start=pulse.start * NS_TO_US,
+        duration=pulse.duration * NS_TO_US,
+        dac=dac,
+        adc=adc,
+        shape=None,
+        name=pulse.serial,
+        type=pulse_type,
```

Comment:
Here you're passing a lot of information about the `pulse`, I wonder if it's possible to just pass the entire `pulse` object (and a few other args), and do all the processing on that side.

Even better: since the `class` and its `__init__` are in the server, and we don't want it to depend on Qibolab, you could do all of this in a separate function.

Reply:
I didn't really understood this comment (and a bit the next one)... it seems to me that you are suggesting what this function already is: a separate function to convert Pulse to rfsoc.Pulse. What would it be the difference?